### PR TITLE
Add fuzzing test for JSON reader

### LINKED ISSUE #4138: [FEA] JSON: fuzz testing
After we have a working JSON reader, #4135 and further extended in #4136, we should try to do some fuzz testing on this.  There are already a lot of corner cases that don't appear to be JSON spec compatible https://github.com/NVIDIA/spark-rapids/issues/9#issuecomment-959803463 and we should look at doing fuzz testing to see what other types of issues we might come up with.

### PR DESCRIPTION
**Pull request by HaoYang670**
Add fuzz tests for JSON reading.
close #4138

This PR contains 3 parts:
1. Schema generator: randomly generate a JSON schema described in PySpark's DataType
2. JSON generator: randomly generator a JSON string based on the given schema
3. test_json_read_fuzz: test the behavior of CPU and GPU on reading a random json file (marked Xfail, so it will not impact the CI job)

Data Type supported so far:
1. primitive types: int, long, float, double, string
2. nested types: struct and array

Plan to support in the future:
1. primitive types: date, timestamp
2. nested types: map

### CODE REVIEW

**Comment by HaoYang670** [comment]
build

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (abellina) [inline integration_tests/src/main/python/json_fuzzing_test.py:397]: should there be at least a second test that is always expected to succeed?

Naively I am also not sure why this test would be expected to fail. I guess the other point to make, would be to add a github issue link on what is failing. Perhaps that's well known to others, but it isn't obvious from this change to me.

Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py:397]: The aim of this fuzzy test is to help find more corner cases in JSON reading. Link: https://github.com/NVIDIA/spark-rapids/issues/4821

Initially, my thought was that it would spoil our CI tests if a corner case was found. So I marked it as `xfail`. However, after thinking carefully about your feedback, I remove the `xfail`. Maybe we should let the CI fail when we find a new corner case, although the possibility is low.

**Comment by HaoYang670** [comment]
build

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (andygrove) [inline integration_tests/src/main/python/json_fuzzing_test.py:25]: We should include a wider range of characters in the generated strings. Spark supports parsing dates and times from JSON strings so we should include characters such as digits and symbols such as ':', ".", and '-'. We may want to also have specific fuzzer code for generating dates and timestamps as strings.

Comment (andygrove) [reply integration_tests/src/main/python/json_fuzzing_test.py:25]: We should also include whitespace characters such as `\n`, `\r`, `\t`, and ' '. Perhaps we can just use the full ASCII character set.

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (andygrove) [inline integration_tests/src/main/python/json_fuzzing_test.py:26]: Are these tests deterministic, using a fixed seed?

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py:25]: @andygrove  The current generator just support generating integer, fp, and string. In the future, I will add more data types such as date, timestamp and map.

Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: > We should also include whitespace characters such as `\n`, `\r`, `\t`, and ' '. Perhaps we can just use the full ASCII character set.

Done

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: Done

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: Done

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

**Comment by pxLi** [comment]
please check the failed log before re-trigger,
```
Error: 3-25T06:40:13.649Z] [ERROR] Failed to execute goal org.apache.rat:apache-rat-plugin:0.13:check (default) on project rapids-4-spark-parent: Too many files with unapproved license: 1 See RAT report in: /home/jenkins/agent/workspace/jenkins-rapids_premerge-github-4260/target/spark24X/rat.txt -> [Help 1]
```

exclude related json filed at https://github.com/NVIDIA/spark-rapids/blob/branch-22.04/pom.xml#L1154-L1178 to pass the license test

**Comment by HaoYang670** [comment]
build

**Comment by HaoYang670** [comment]
build

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (revans2) [inline integration_tests/src/main/python/json_fuzzing_test.py:398]: To be clear fuzz testing is intended to find issues and should not be running by default as a part of integration tests. It should be something that we can enable and run for a configurable number of iterations to find problems. Once we find the problem we need to have a way to debug/reproduce this issue. Then we can file a follow on issue/bug to fix it. 

I would much rather have us skip this test by default unless a flag is set to enable fuzz testing, preferably with a number of iterations.  With xfail we want to get to the point where if a test passes when we expect it to fail then it will be marked as a failure, `xfail_strict=true`. Because this is totally random I really don't want to mark it as xfail.

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (revans2) [inline integration_tests/src/main/python/json_fuzzing_test.py:400]: How do we reproduce/debug an error when it happens? Everything is random and based off of a regular random. Typically in python we would want to create an instance of Random with a seed, and then pass it everywhere so the results can be reproduced. We do this in data_gen already.

https://github.com/NVIDIA/spark-rapids/blob/00630537365c034186fa9031694edb503e4341a6/integration_tests/src/main/python/data_gen.py#L671

I am fine if we randomly generate a seed based off of a timestamp or something like that. But at a minimum we need a way to log what that seed was so if/when it does fail we can debug things.  With xdist this is not super simple, so I would suggest that we catch all Exceptions and then wrap them with something new that includes the seed that triggered the problem.

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: Done. Fuzz tests are skipped by defualt.
We can enable fuzz tests and store the tested data by running
```
> bash run_pyspark_from_build.sh -k test_json_read_fuzz --fuzz_test --debug_tmp_path
```

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: Add the flag `--debug_tmp_path` can save the test data

File: integration_tests/src/main/python/json_fuzzing_test.py
Comment (HaoYang670) [reply integration_tests/src/main/python/json_fuzzing_test.py]: Thank you @andygrove . Really nice reminder! I will add these and other types supported by Spark in following PRs.

**Comment by HaoYang670** [comment]
build

File: integration_tests/README.md
Comment (revans2) [inline integration_tests/README.md:384]: nit: You have "th" instead of "the"

The tests can be enabled by appending the option `--fuzz_test` to the command.

File: integration_tests/README.md
Comment (revans2) [inline integration_tests/README.md:388]: Could we combine these two together? So if you pass in `--fuzz_test_debug_path` or something like that the tests are enabled? To be clear I am fine with keeping the debug_tmp_path as it is. If this is really hard you don't need to do it.

File: integration_tests/src/main/python/json_fuzz_test.py
Comment (revans2) [inline integration_tests/src/main/python/json_fuzz_test.py:17]: Why do we need this? We are not doing CGI are we?

File: integration_tests/README.md
Comment (HaoYang670) [reply integration_tests/README.md:384]: Done!

**Review by nartal1 — APPROVED** [review]
LGTM but would be good to get review from others who had reviewed earlier.

**Comment by HaoYang670** [comment]
build

**Comment by sameerz** [comment]
Does this close issue https://github.com/NVIDIA/spark-rapids/issues/4138 ?

**Comment by HaoYang670** [comment]
> Does this close issue #4138 ?

Yes. Have updated.

**Comment by HaoYang670** [comment]
Hi @revans2. Do you think we could merge this PR or not?

**Comment by revans2** [comment]
@HaoYang670 I am very busy with other things. Please don't let me block this from going in. If others have approved it then we can merge it in and if there are issues we can iterate on them as we find them.

**Review by andygrove — APPROVED** [review]

**Pull request merged**
